### PR TITLE
feat(api): Generate a blake2b hash of a given binary buffer.

### LINKED ISSUE #98: 🛠️ [TASK] : Hashing utility API: Hermes Implementation (Rust Module)

### PR DESCRIPTION
**Pull request by bkioshn**
# Description

Generate a blake2b hash of a given binary buffer. 
Related function include
- `blake2b`
- `blake2bmac`

## Related Issue(s)

Closes #98 

## Description of Changes

Reference to blake2b can be found [rfc](https://datatracker.ietf.org/doc/html/rfc7693) and [doc](https://www.blake2.net/blake2.pdf)

Fix how blake2 is implemented, since when blake2 is keyed it changes to MAC (Message Authentication Code) mode.
- Divide blake2 into 2 functions
    - blake2b: `blake2b` and `blake2bmac`
    - blake2s: `blake2s` and `blake2smac`
- `blake2` without `key` will only accept bytes of `buf` and optional `outlen` which is the output length of the hash (eg. if `outlen` = 32 the hash function is equal to `blake2b-256`)
-  `blake2mac` accepts bytes of `buf`, optional `outlen`, `key`, optional `salt`, and optional `personal`
**If `outlen` is not specified, set to 64 which is blake2b-512**
- Handle all possible errors
- Create rust unit test for `blake2b` and `blake2bmac`

## Please confirm the following checks

* [x] My code follows the style guidelines of this project
* [x] I have performed a self-review of my code
* [x] I have commented my code, particularly in hard-to-understand areas
* [x] I have made corresponding changes to the documentation
* [x] My changes generate no new warnings
* [x] I have added tests that prove my fix is effective or that my feature works
* [x] New and existing unit tests pass locally with my changes
* [x] Any dependent changes have been merged and published in downstream module

### CODE REVIEW

File: hermes/bin/Cargo.toml [hunk 1]
```diff
@@ -16,3 +16,5 @@ workspace = true
 wasmtime = { workspace = true, features = ["component-model"] }
 rusty_ulid = { workspace = true }
 anyhow = { workspace = true }
+blake2.workspace = { workspace = true }
```
Comment (Mr-Leshiy) [inline hermes/bin/Cargo.toml:19]:
```suggestion
blake2 = { workspace = true }
```

File: hermes/bin/Cargo.toml [hunk 2]
```diff
@@ -16,3 +16,5 @@ workspace = true
...
+hex-literal.workspace = { workspace = true }
```
Comment (Mr-Leshiy) [inline hermes/bin/Cargo.toml:20]:
```suggestion
hex-literal = { workspace = true }
```

File: hermes/bin/src/runtime/host/hermes/hash/mod.rs [hunk 3]
```diff
@@ -30,8 +34,13 @@ impl Host for HermesState {
     fn blake2b(
         &mut self, buf: Bstr, outlen: Option<u8>, key: Option<Bstr>,
     ) -> wasmtime::Result<Result<Bstr, Errno>> {
-        // WIP
-        todo!()
+        // Recheck this implementation might need to use Blake2sMac256 for adding the key
+        let mut v = Blake2bVarCore::new_with_params(salt, persona, key_size, output_size);
```
Comment (Mr-Leshiy) [inline hermes/bin/src/runtime/host/hermes/hash/mod.rs:38]: I would rather move all this code under some free function `blake2b(...)` and call it inside `HermesState::blake2b`.
So the unit test would be much simpler and will cost for us nothing, you will not need to setup a `Hermes` instance,  wasm module etc.

And to proper test how it is working under the WASM context, we will would rather write integration tests, but it is another story.

**Review by stevenj — COMMENTED** [review]
I know i suggested this blake2 crate, but i see the problems it causes by not being a good match for our requirements.
Have a look at https://crates.io/crates/blake2b_simd
Which might be a better fit.

File: hermes/bin/src/runtime/host/hermes/hash/mod.rs [hunk 4]
```diff
@@ -18,19 +28,113 @@ impl Stateful for State {
...
+    fn blake2bmac(
+        &mut self, buf: Bstr, outlen: Option<u8>, key: Bstr, salt: Option<Bstr>,
+        persona: Option<Bstr>,
     ) -> wasmtime::Result<Result<Bstr, Errno>> {
-        todo!()
+        let hash = blake2bmac_impl(&buf, outlen, &key, salt, persona);
+        match hash {
+            Ok(hash) => Ok(Ok(hash)),
+            Err(err) => Err(err.into()),
+        }
```
Comment (apskhem) [inline hermes/bin/src/runtime/host/hermes/hash/mod.rs:137]:
```suggestion
        blake2bmac_impl(&buf, outlen, &key, salt, persona)
            .map(Ok)
            .map_err(Into::into)
```

File: hermes/bin/src/runtime/host/hermes/hash/mod.rs [hunk 5]
```diff
@@ -18,19 +28,113 @@ impl Stateful for State {
...
+    // TODO - I don't think it is possible to pass outlen to Blake2bMac
+    if outlen == 64 {
+        return blake2bmac::<U64>(buf, key, &salt, &persona);
+    } else if outlen == 32 {
+        return blake2bmac::<U32>(buf, key, &salt, &persona);
+    } else if outlen == 48 {
+        return blake2bmac::<U48>(buf, key, &salt, &persona);
+    } else if outlen == 20 {
+        return blake2bmac::<U20>(buf, key, &salt, &persona);
+    }
```
[the remark this answers is not in the document]
Comment (bkioshn) [reply hermes/bin/src/runtime/host/hermes/hash/mod.rs:101]: Pattern matching doesn't work here. Will change the crate

File: hermes/bin/src/runtime/host/hermes/hash/mod.rs [hunk 6]
```diff
@@ -18,19 +28,113 @@ impl Stateful for State {
...
 
     /// Hash a binary buffer with `BLAKE2b`
-    fn blake2b(
-        &mut self, buf: Bstr, outlen: Option<u8>, key: Option<Bstr>,
+    fn blake2b(&mut self, buf: Bstr, outlen: Option<u8>) -> wasmtime::Result<Result<Bstr, Errno>> {
+        let hash = blake2b_impl(&buf, outlen);
+        match hash {
+            Ok(hash) => Ok(Ok(hash)),
+            Err(err) => Err(err.into()),
+        }
```
[the remark this answers is not in the document]
Comment (bkioshn) [reply hermes/bin/src/runtime/host/hermes/hash/mod.rs]: Thank you, much cleaner

File: hermes/bin/src/runtime/host/hermes/hash/blake2b.rs [hunk 7]
```diff
@@ -0,0 +1,190 @@
+//! Implementation of blake2b and blake2bmac hash functions.
+
+use blake2b_simd::Params;
+
+use crate::runtime::extensions::bindings::hermes::{binary::api::Bstr, hash::api::Errno};
+
+/// Implementation of blake2b given a buffer and outlen.
+pub(crate) fn blake2b_impl(buf: &Bstr, outlen: Option<u8>) -> Result<Bstr, Errno> {
+    // Default to 64 bytes Blake2b-512
+    let outlen = outlen.unwrap_or(64) as usize;
```
Comment (Mr-Leshiy) [inline hermes/bin/src/runtime/host/hermes/hash/blake2b.rs:10]:
```suggestion
    let outlen: usize = outlen.unwrap_or(64).into();
```

File: hermes/bin/src/runtime/host/hermes/hash/blake2b.rs [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+
+    return Ok(Bstr::from(hash.as_bytes()));
+}
+
+/// Implementation of blake2bmac given a buffer, outlen, key, salt, and persona.
+pub(crate) fn blake2bmac_impl(
+    buf: &Bstr, outlen: Option<u8>, key: &Bstr, salt: Option<Bstr>, personal: Option<Bstr>,
+) -> Result<Bstr, Errno> {
+    // Default to 64 bytes Blake2b-512
+    let outlen = outlen.unwrap_or(64) as usize;
```
Comment (Mr-Leshiy) [inline hermes/bin/src/runtime/host/hermes/hash/blake2b.rs:32]:
```suggestion
    let outlen: usize = outlen.unwrap_or(64).into();
```

File: hermes/bin/src/runtime/host/hermes/hash/blake2b.rs [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+    let hash = Params::new()
+        .hash_length(outlen)
+        .key(key)
+        .salt(&salt)
+        .personal(&personal)
+        .to_state()
+        .update(buf)
+        .finalize();
+
+    return Ok(Bstr::from(hash.as_bytes()));
```
Comment (Mr-Leshiy) [inline hermes/bin/src/runtime/host/hermes/hash/blake2b.rs:67]:
```suggestion
    return Ok(hash.as_bytes().into());
```

File: hermes/bin/src/runtime/host/hermes/hash/blake2b.rs [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+    } else if outlen > 64 {
+        return Err(Errno::HashTooBig);
+    }
+    let hash = Params::new()
+        .hash_length(outlen)
+        .to_state()
+        .update(buf)
+        .finalize();
+
+    return Ok(Bstr::from(hash.as_bytes()));
```
Comment (Mr-Leshiy) [inline hermes/bin/src/runtime/host/hermes/hash/blake2b.rs:24]:
```suggestion
    return Ok(hash.as_bytes().into());
```

File: hermes/bin/src/runtime/host/hermes/hash/blake2b.rs [hunk 11]
```diff
@@ -0,0 +1,190 @@
+//! Implementation of blake2b and blake2bmac hash functions.
+
+use blake2b_simd::Params;
+
+use crate::runtime::extensions::bindings::hermes::{binary::api::Bstr, hash::api::Errno};
+
+/// Implementation of blake2b given a buffer and outlen.
```
Comment (Mr-Leshiy) [inline hermes/bin/src/runtime/host/hermes/hash/blake2b.rs:7]: Describe as a part of the documentation what probable errors could occur.

File: hermes/bin/src/runtime/host/hermes/hash/blake2b.rs [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+    let hash = Params::new()
+        .hash_length(outlen)
+        .to_state()
+        .update(buf)
+        .finalize();
+
+    return Ok(Bstr::from(hash.as_bytes()));
+}
+
+/// Implementation of blake2bmac given a buffer, outlen, key, salt, and persona.
```
Comment (Mr-Leshiy) [inline hermes/bin/src/runtime/host/hermes/hash/blake2b.rs:27]: Describe as a part of the documentation what probable errors could occur.

File: hermes/bin/src/runtime/host/hermes/hash/mod.rs [hunk 13]
```diff
@@ -23,18 +25,36 @@ impl Stateful for State {
...
+    fn blake2smac(
+        &mut self, _buf: Bstr, _outlen: Option<u8>, _key: Bstr, _salt: Option<Bstr>,
+        _persona: Option<Bstr>,
     ) -> wasmtime::Result<Result<Bstr, Errno>> {
         todo!()
     }
 
+    /// Hash a binary buffer with `BLAKE2b`
+    fn blake2b(&mut self, buf: Bstr, outlen: Option<u8>) -> wasmtime::Result<Result<Bstr, Errno>> {
+        blake2b::blake2b_impl(&buf, outlen)
```
Comment (Mr-Leshiy) [inline hermes/bin/src/runtime/host/hermes/hash/mod.rs:43]: I am not sure that this is a correct conversion which you are doing here.
It seems like you are converting a `Errno` to the `wasmtime::Error` which is not correct for this place.
I think here you should return the following.
```
Ok( blake2b::blake2b_impl(&buf, outlen))
```

File: hermes/bin/src/runtime/host/hermes/hash/mod.rs [hunk 14]
```diff
@@ -23,18 +25,36 @@ impl Stateful for State {
...
+            .map(Ok)
+            .map_err(Into::into)
+    }
+
+    /// Hash a binary buffer with `BLAKE2b` with `MAC` (Message Authentication Code) mode
+    fn blake2bmac(
+        &mut self, buf: Bstr, outlen: Option<u8>, key: Bstr, salt: Option<Bstr>,
+        personal: Option<Bstr>,
+    ) -> wasmtime::Result<Result<Bstr, Errno>> {
+        blake2b::blake2bmac_impl(&buf, outlen, &key, salt, personal)
```
Comment (Mr-Leshiy) [inline hermes/bin/src/runtime/host/hermes/hash/mod.rs:53]: Same as above

**Review by Mr-Leshiy — CHANGES_REQUESTED** [review]

File: hermes/bin/src/runtime/host/hermes/hash/blake2b.rs [hunk 15]
```diff
@@ -0,0 +1,224 @@
...
+        .to_state()
+        .update(buf)
+        .finalize();
+
+    return Ok(hash.as_bytes().into());
+}
+
+/// Implementation of blake2bmac given a buffer, outlen, key, salt, and persona.
+///
+/// **Parameters**
```
Comment (Mr-Leshiy) [inline hermes/bin/src/runtime/host/hermes/hash/blake2b.rs:41]: Update the format to the following:
```
# Arguments
- ...

# Errors
- ...
```
It is a common way for Rust how to describe such sections https://doc.rust-lang.org/book/ch14-02-publishing-to-crates-io.html#commonly-used-sections.

File: hermes/bin/src/runtime/host/hermes/hash/blake2b.rs [hunk 16]
```diff
@@ -0,0 +1,224 @@
+//! Implementation of blake2b and blake2bmac hash functions.
+
+use blake2b_simd::Params;
+
+use crate::runtime::extensions::bindings::hermes::{binary::api::Bstr, hash::api::Errno};
+
+/// Implementation of blake2b.
+///
+/// **Parameters**
```
Comment (Mr-Leshiy) [inline hermes/bin/src/runtime/host/hermes/hash/blake2b.rs:9]: same

**Review by Mr-Leshiy — DISMISSED** [review]
Besides few comments LGTM

**Review by stevenj — APPROVED** [review]
Looks great.

**Pull request merged**